feat: -incsearch option for user-defined commands

mg979 opened the pull request:
So that they can also highlight their argument, as do :substitute,
:global, :vimgrep, etc.

mg979 committed:
feat: -incsearch option for user-defined commands

So that they can also highlight their argument, as do :substitute,
:global, :vimgrep, etc.

mg979 committed:
Make -incsearch require -range

Not strictly necessary but should prevent abuse, also considering that
range defaults to current line only, so that having a buffer range must
be explicit.

bfredl commented:
nice addition. also should be supported in `nvim_add_user_command` (reuse the pattern for existing booleans).

mg979 committed:
Can be set with nvim_add_user_command

mg979 commented:
Thanks! I think there is a problem in `nvim_add_user_command` in that it accepts non-alphanumeric names for commands, by mistake I made a command named `Find!` and it accepted it, it appeared in the completion list for commands, but then it said `invalid command`.

Also it doesn't check if a command with the same name exists, then it says `add ! to replace it` but it's out of context because it's the error message for the `command` ex command (what led me to add `Find!` in the first place.

mg979 commented:
Also I'm seeing that it advances the pointer twice to find the whitespace character, once in the helper and once in the main function, should I fix it?

mg979 committed:
Advance the pointer to first whitespace only once

mg979 committed:
Ensure there's actually some argument

mg979 commented:
Also the whole thing is a bit expensive because it must loop over all user commands every time, it would be better to have dedicated lists of user commands that can do incsearch, then loop over them exclusively. Is there some linked list type that nvim uses? I could only find `garray_T`.

gpanders commented on src/nvim/ex_getln.c:
```diff
@@ -349,6 +349,43 @@ static void get_healthcheck_cb(char_u *path, void *cookie)
 
+// Return an user command when it can do incsearch highlighting, NULL otherwise.
+// Will also advance the pointer to the end of the command name, if appropriate.
+static ucmd_T *usrcmd_can_incsearch(char_u *cmd_name, char_u **p)
```
Minor nit but for the sake of consistency:
```suggestion
static ucmd_T *uc_can_incsearch(char *cmd_name, char **p)
```

We also try to avoid adding new usages of `char_u`, so for new functions just use `char` and cast where necessary.

mg979 replied:
Ok

mg979 committed:
Use char, not char_u

mg979 committed:
Different take that doesn't pass pointers

Also more efficient when it's an user command, but
do_incsearch_highlighting is more complex now.

mg979 commented:
For me it's either https://github.com/neovim/neovim/pull/16831/commits/74d3101e24046a14ed098a19bdd6a4144045efe2 or https://github.com/neovim/neovim/pull/16831/commits/6d1210eb4bcd1464c5833315c549cef65e80ce3d.

The first one is clearer and simpler, but less efficient. `char_u` should be still fixed there.

mg979 committed:
Remove also_uc label

mg979 committed:
Move attribute check

So that it happens when all attributes have been already scanned,
otherwise

  command -incsearch -range ...

wouldn't work, but

  command -range -incsearch ...

would work.
